README.md
![封面](https://raw.githubusercontent.com/looly/hutool/master/docs/resources/hutool.jpg)

[![Version](https://img.shields.io/badge/version-3.0.7-brightgreen.svg)](http://search.maven.org/#search%7Cga%7C1%7Ca%3A%22hutool-all%22)
[![License](http://img.shields.io/:license-apache-blue.svg)](http://www.apache.org/licenses/LICENSE-2.0.html)
[![JDK 1.7](https://img.shields.io/badge/JDK-1.7-green.svg "JDK 1.7")]()

## 全新项目主页：[http://www.hutool.cn/](http://www.hutool.cn/)

## Hutool交流QQ群：== [537116831](http://shang.qq.com/wpa/qunwpa?idkey=382bb37ce779c11da77577f69d92d5171b340e3e7343d5ae0521f237c82c7810) ==

## 功能
一个Java基础工具类，对文件、流、加密解密、转码、正则、线程、XML等JDK方法进行封装，组成各种Util工具类，同时提供以下组件：
* 布隆过滤
* 缓存
* 克隆接口
* 类型转换
* 日期处理
* 数据库ORM（基于ActiveRecord思想）
* 基于DFA有限自动机的多个关键字查找
* HTTP客户端
* IO和文件
* 有用的一些数据结构
* 日志
* 反射代理类的简化（AOP切面实现）
* Setting（一种扩展Properties的配置文件）
* System（JVM和系统信息等）
* WatchService的封装（文件变动监控）
* XXXUtil各种有用的工具类

## 简介
[Hutool](https://github.com/looly/hutool)是一个工具包，我日常写项目的一些积累，希望你看了之后会有所启发或者能给你工作中带来帮助。这个工具包叫做Hutool也有“糊涂”之意，意为很多时候我们并不需要关注细节，专注业务。

## 设计哲学

### 1. 方法优先于对象
在工具类中，往往以静态方法为主。方法集中在一个类中，配合IDE查找使用起来是十分便利的。于是Hutool将JDK中许多的类总结抽象为一个方法，这一原则使用最多的就是流的相关方法，这些方法很好的隐藏了XXXInputStream、XXXReader等的复杂性。

### 2. 自动识别优于用户定义
其实很多时候，有些参数、设置等是没有必要我们自己传入的，完全可以靠逻辑判断自动完成。一个方法很多时候明明只需要传3个参数，我们非要传4个，这多出的一个参数本身就是代码的一种冗余。

这一原则在Hutool的各个角落都有所体现，尤为明显的比如log模块。构建日志对象的时候，很明显类名可以动态获取，何必让使用者再传入呢？再比如在db模块的数据库配置中，数据库驱动命名完全可以根据连接字符串判断出来，何必要让用户传入？这些问题的在Hutool中都有非常好的封装，而这一原则也渐渐变成Hutool哲学的一部分。

### 3. 便捷性与灵活性并存
所谓便捷性，就是我们在调用一个方法的时候参数要尽量少，只传必要参数即可，非必要参数使用默认值即可（想想一个方法一堆参数的时候，调用者晕头转向不知所云）。

所谓灵活性正好与便捷性相反，要让一个方法的参数尽量多，为用户灵活的操作方法提供最大可能性。

这两个原则看似矛盾，其实只是针对不同场景设定的而已，缺一不可。便捷性强调拿来即用，为快速开发提供可能；灵活性强调最大限度调优，为性能调优和扩展提供便利。

这一原则在针对编码问题上体现尤为突出，我们的大部分方法都是默认“UTF-8”编码的，这也是我们推荐的编码方式，推荐大部分项目使用的编码。但是一旦有遗留项目使用了类似“GBK”等编码，没关系，我们提供在相关方法中提供Charset对象参数，可以自定义编码。这样使用这一原则就兼顾了各种项目的情况。

### 4. 适配与兼容
在Hutool中，适配器模式运用特别广泛，log模块适配主流各大框架，db模块适配主流各种连接池和关系数据库。这种适配一是提高灵活性，二是可以很好的兼容各大框架，让Hutool可以在各种复杂项目环境中生存的很好。

适配兼容产生的另一个原则是：**你有我配，你无我有**。说白了就是：如果你项目中有这个框架，我可以完美适配，如果你没有引入任何框架，Hutool自身实现了一些逻辑可以很好的工作。

### 5. 可选依赖原则
在Java项目中依赖常常是个头疼的问题，不同的框架强依赖另一些框架或包，虽然Maven可以很好的处理冲突问题，但是项目底下满满的依赖jar包，是不是无形中拖慢了项目，也增加了复杂性和不确定性？而很多时候，我们是不是只是为了用一个小小的方法，就要引入一个第三方包，谁喜欢这样臃肿的项目？

Hutool中也会有一些依赖，但是全部都是**optional**的，在使用中不会关联依赖，而这些依赖只有在使用者使用到时才会调用，这时可能会报ClassNotFoundException，不用担心，我们自己引入即可。为什么要这样做呢？以VelocityUtil这个工具类为例，使用Velocity的人占比极少，我们不能为了这些用户而强引入Velocity包，而使用这个工具类的人应该明白，我们应该自己引入这个包。

而更多时候，我们需要用到某个方法时，我的做法是将方法拷贝到项目中（Hutool中的方法正在不断积累），类似于Apache Commons中的方法，Hutool中基本都有取代方法，完全不必要引入。

可选依赖原则让我们的项目更加精简，问题也更容易排查。

### 6. 无侵入原则
Hutool始终是一个工具类而不是框架，这意味着它对项目的侵入几乎为零，每个方法都是可被代替的，甚至整个Hutool也是可被替换的。这种无侵入性，让使用者可以更加放心的在项目中引入，也保证了与其它框架完美的兼容。

## 安装
### Maven
在项目的pom.xml的dependencies中加入以下内容:

```xml
<dependency>
    <groupId>com.xiaoleilu</groupId>
    <artifactId>hutool-all</artifactId>
    <version>3.0.7</version>
</dependency>
```

### Gradle
```
compile 'com.xiaoleilu:hutool-all:3.0.7'
```

注：工具包的**版本**可以通过 [http://search.maven.org/](http://search.maven.org/) 搜索`hutool-all`找到项目。

### 非Maven项目
可以从[http://search.maven.org/](http://search.maven.org/) 搜索`hutool-all`找到项目，点击对应版本，下面是相应的Jar包，导入即可使用。

[http://search.maven.org/#search%7Cgav%7C1%7Cg%3A%22com.xiaoleilu%22%20AND%20a%3A%22hutool-all%22](http://search.maven.org/#search%7Cgav%7C1%7Cg%3A%22com.xiaoleilu%22%20AND%20a%3A%22hutool-all%22)

点击链接后点选择对应版本，点击列表尾部的“Download”下载jar、API文档、源码

## 文档请移步 

[Hutool Wiki@OSC](http://hutool.mydoc.io/)

## 版本变更

[CHANGELOG.md](https://github.com/looly/hutool/blob/master/CHANGELOG.md)


CHANGELOG.md
# Changelog

## 3.0.7

### 新特性
* 增加NetUtil.getLocalhostStr（感谢【帝都】-宁静）
* ArrayUtil和CollectionUtil增加hasNull方法
* 添加PatternPool用于存放常用编译好的正则表达式
* 完善Assert类，可变消息参数以及规范注释
* 添加NetUtil.getMacAddress和NetUtil.getLocalMacAddress用于获取MAC地址
* StrUtil.repeat方法优化
* 增加comparator包，补充ComparableComparator，ComparatorChain，ReverseComparator
* 增加BeanResolver，同时增加BeanUtil.getProperty方法
* 增加WeakCache
* HttpRequest中增加cookie方法用于覆盖默认的Cookie行为，自定义Cookie值（感谢@质量过关）
* getPropertyDescriptor和getPropertyDescriptorMap增加缓存支持
* 添加DynaBean，反射方式对Bean做get和set操作
* ArrayUtil中数组参数变为可变参数。提供跟多灵活性
* StrUtil增加更多实用方法
* DateUtil.date方法支持java.util.Date参数；DateTime增加toTimestamp和toSqlDate两个方法
* FileUtil增加getResourceUrl方法

### Bug修复
* Validator.isPlateNmber 拼写修复为isPlateNumber（感谢：飞天奔月）
* 修复DbUtil.getTableMeta获取主键出错问题
* 修复ConverterRegistry中默认Date对象错误问题（issue#22@github）
* 修复NumberUtil中方法非static的问题（感谢[霾都] QQ小冰）
* 修复FileUtil.equals方法，实现改为JDK7实现。
* 修复FileUtil.copy 方法判断错误问题（issue#24@github）
* 修复Bean中Date对象转JSON格式问题（会导致toBean异常）（感谢@【武汉】徐元程）
* 修复Cache模块并发修改Map会导致的异常（感谢@【北京】liyong）

-------------------------------------------------------------------------------------------------------------

## 3.0.6

### 新特性
* ThreadUtil增加newExecutorByBlockingCoefficient方法（感谢：@【北京】仓山有井名为空）
* 解决LogFactory中频繁创建Log对象造成的性能问题（issue#19@osc）
* 解决LFUCache中负载因子导致的扩容rehash问题（issue#18@osc）
* FileUtil中IOException全部包装为IORuntimeException
* 针对JDK8文档注释修复
* 添加SystemClock
* 增加ArrayUtil.copy方法
* 增加FtpUtil
* 增加FileUtil.clean方法，用于清空目录
* 增加SqlRunner.create()方法，使用默认连接
* 扩充Entity类型
* Setting添加Profile功能支持（感谢@长沙-渔泯小镇）
* ZipUtil增加对输入流和字符串压缩的支持
* Validator增加中国车牌号码
* 新增StrUtil.cut方法，等分切割
* 增加HttpUtil.download方法对https的支持

### Bug修复
* 修复AbstractCache.onRemove回调方法在get和pruneCache时不被触发问题（issue#18@osc）
* 技术债务修复：空指针等问题修复
* 修复RandomUtil.randomEleSet方法获取随机个数出错问题。（感谢@【北京】Sych）
* NumberUtil的isInteger和isDouble方法非静态问题修复（感谢@【广州】流行の云）
* Base64的URL Safe模式等号没有替换问题。（issue#20@osc）
* 修复Convert.strToUnicode转换后位数不足问题
* 修复针对StrUtil.cleanBlank和Convert.toDBC中不间断空白符无法去除问题
* 修复db模块对表名的包装策略
* 修复BeanToMap和JSON中toMap导致的循环引用
* 修复在自动侦测数据源的时候重复加载配置文件问题
* 修复针对JSON转Bean中数组参数转换失败的问题，同时添加ArrayUtil.cast方法用于强制转换数组类型
* 修复NumberUtil.isInteger方法中对空串的判断(#22@osc)
* 修复dfa模块匹配时正文中存在停顿词（特殊字符）时导致匹配失效问题（感谢@talent-aio）

-------------------------------------------------------------------------------------------------------------

## 3.0.5

### 新特性
* ReUtil.replaceAll扩充支持Pattern对象
* 优化log模块中自动选择日志实现机制
* 针对不同连接池添加样例配置文件
* AsymmetricCriptor类增加构造方法，可以传入Base64编码的私钥和公钥
* SecureUtil增加RSA和DSA快速创建工具方法
* Hutool-log支持tinylog
* 添加GlobalThreadPool，将公共线程池从ThreadUtil中剥离出来
* 扩充Base32中的方法

### Bug修复
* db模块报错问题修复（感谢@尘风了了）
* DateUtil.month方法注释修复 (感谢@〓下页)
* 对db模块的数据源部分的配置做修复
* Base64传入null导致空指针异常修复（解决方法为返回null）
* crypto模块类名存在拼写错误：修正SymmetricCriptor -> SymmetricCrypto，AsymmetricCriptor -> AsymmetricCrypto
* FileCache中初始化capacity为0问题。（issue#17@osc）

-------------------------------------------------------------------------------------------------------------

## 3.0.4

### 新特性
* 新增CollectionUtil.newCopyOnWriteArrayList方法
* 新增IdcardUtil，身份证验证和相关信息提取
* Convert内部修改，避免循环引用
* ArrayUtil.zip方法添加参数isOrder，可选是否有序
* 增加HexUtil.decodeHexStr(String) 感谢（@MaxSherry）pull request #15
* 增加SymmetricCriptor.encryptHex 感谢（@MaxSherry）pull request #15
* JavaInfo增加对JDK1.8的判断 感谢（@MaxSherry）pull request #19
* BeanUtil.copyProperties方法增加参数，可选是否忽略大小写 感谢（@talent-aio）

### Bug修复
* 修复Convert.toBigDecimal和toBigInteger值传为默认值的问题（感谢@〆得不到的永远在骚动）
* 修复AsymmetricCriptor.getKeyByType中判断出错问题，感谢（@姚洪涛）
* 修复Http模块中Proxy设置无效问题，现在Proxy对象在HttpRequest中设置
* Convert中数组转换修复方法，消除歧义
* 修复BeanUtil.fillBean注入时空指针问题
* 修复Aspect中afterException对象传错问题 #17@github 感谢（@970655147）
* 修复CaseInsensitiveMap导致的忽略大小写注入问题
* 解决LongBitMap中强转位丢失问题

-------------------------------------------------------------------------------------------------------------

## 3.0.3

### 新特性
* DigestUtil中增加hmac方法，HMac对象快速构建通道
* HMac类中增加getSecretKey方法，获取密钥

### Bug修复
* 修复Digester.digest方法返回null的问题（issue#16@osc）
* 修复HMac.digest方法返回null的问题
* HttpBase中的header等添加this返回（感谢github@MaxSherry）

-------------------------------------------------------------------------------------------------------------

## 3.0.2

### 新特性
* 增加CaseInsensitiveMap

### Bug修复
* 解决JSONObject.toBean方法无法多层转换的问题
* 解决FileUtil.append非追加而是覆盖问题（感谢 @Andy）

-------------------------------------------------------------------------------------------------------------

## 3.0.1

### 新特性
* 拆分项目
* 增加 ThreadUtil.safeSleep方法
* 增加CollectionUtil.addAllIfNotContains方法
* 增加NumberUtil
* 增加hutool-cron模块
* 增加 ThreadUtil.waitForDie
* 增加DateUtil.betweenDay方法（感谢 @宁静之湖）
* 增加Snowflake算法
* 对Util类加final修饰符，并添加private构造方法
* 增加 ThreadUtil.interupt方法
* Setting中对RuntimeException变为SettingRuntimeException
* [log] LogFactory.setCurrentLogFactory支持传入LogFactory类
* [core] 增加CollectionUtil.getFirst
* [core] NumberUtil中增加binaryToInt、binaryToLong、getBinaryStr
* 增加Mutable类型
* HttpUtil.downFile增加String参数方法
* 修复NumberUtil.round方法传入double某些数字无效问题
* 增加FileUtil.getType方法
* 优化整理JSON部分
* 增加ThreadUtil.getThreads、getMainThread方法
* 增强JarClassLoader
* WatchMonitor增加createAll方法
* 增加NetUtil.getLocalhost
* 优化ClassPathResource错误提示
* 增加 ArrayUtil.toArray方法
* MathUtil中的方法迁移到NumberUtil，去除MathUtil
* 针对Cache模块做包结构调整，扩展FileCache

### Bug修复
* DateUtil中offsite修正为offset
* ClassUtil.invoke方法在执行无法实例化的静态方法时报错问题
* 修复克隆方法对数组支持不足问题
* Convert修复众多bug
* 去掉IoUtil.getReader无用的IO异常抛出
* ImageUtil.pressImage修正值无效问题（感谢@Rainplus）
* 修复cache模块中CacheValuesIterator对象next()方法没有元素时无抛出异常问题

-------------------------------------------------------------------------------------------------------------
-------------------------------------------------------------------------------------------------------------

## 2.16.2

### 新特性
* SecureUtil对常用算法静态函数化封装
* 丰富Sql查询种类
* 对db中的Session和SqlRunner中的方法抽象到AbstractRunner中
* 增加CollectionUtil.split方法
* 增加 BOMInputStream.java用于读取带BOM头的文件，同时FileUtil提供getBOMInputStream方法
* 扩充CollectionUtil
* 扩充FileUtil，增加针对JDK7新特性的文件拷贝方法
* 对File的读写抽象为FileReader和FileWriter
* 添加Season，对季度的Enum支持
* HexUtil增加重载方法

hutool-core/src/main/java/com/xiaoleilu/hutool/lang/Base64.java
package com.xiaoleilu.hutool.lang;

import java.nio.charset.Charset;

import com.xiaoleilu.hutool.util.CharsetUtil;
import com.xiaoleilu.hutool.util.StrUtil;

/**
 * Base64工具类，提供Base64的编码和解码方案<br>
 * base64编码是用64（2的6次方）个ASCII字符来表示256（2的8次方）个ASCII字符，也就是三位二进制数组经过编码后变为四位的ASCII字符显示，长度比原来增加1/3。
 * 
 * @author Looly
 *
 */
public final class Base64 {
	
	private Base64() {}

	/** 标准编码表 */
	private static final byte[] STANDARD_ENCODE_TABLE = { 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '+', '/' };
	/** URL安全的编码表，将 + 和 / 替换为 - 和 _ */
	private static final byte[] URL_SAFE_ENCODE_TABLE = { 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '-', '_' };
//	private static final byte STANDARD_DECODE_TABLE[] = { -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 62, -1, -1, -1, 63, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, -1, -1, -1, -1, -1, -1, -1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, -1, -1, -1, -1, -1, -1, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51 };
//	private static final byte URL_SAFE_DECODE_TABLE[] =   { -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 0, 1, 2, 3, 4, 5, 6, 7, 8, -1, 62, 9, 10, 11, -1, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 12, 13, 14, -1, 15, 63, 16, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 17, -1, 18, 19, 21, 20, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 22, 23, 24, 25 };
	/** Base64解码表 */
	private static final byte[] DECODE_TABLE = { -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 62, -1, 62, -1, 63, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, -1, -1, -1, -1, -1, -1, -1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, -1, -1, -1, -1, 63, -1, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51 };

	// -------------------------------------------------------------------- encode
	/**
	 * 编码为Base64，非URL安全的
	 * 
	 * @param arr 被编码的数组
	 * @param lineSep 在76个char之后是CRLF还是EOF
	 * @return 编码后的bytes
	 */
	public static byte[] encode(byte[] arr, boolean lineSep) {
		return encode(arr, lineSep, false);
	}
	
	/**
	 * 编码为Base64，URL安全的
	 * 
	 * @param arr 被编码的数组
	 * @param lineSep 在76个char之后是CRLF还是EOF
	 * @return 编码后的bytes
	 * @since 3.0.6
	 */
	public static byte[] encodeUrlSafe(byte[] arr, boolean lineSep) {
		return encode(arr, lineSep, true);
	}

	/**
	 * base64编码
	 * 
	 * @param source 被编码的base64字符串
	 * @return 被加密后的字符串
	 */
	public static String encode(String source) {
		return encode(source, CharsetUtil.UTF_8);
	}
	
	/**
	 * base64编码，URL安全
	 * 
	 * @param source 被编码的base64字符串
	 * @return 被加密后的字符串
	 * @since 3.0.6
	 */
	public static String encodeUrlSafe(String source) {
		return encodeUrlSafe(source, CharsetUtil.UTF_8);
	}

	/**
	 * base64编码
	 * 
	 * @param source 被编码的base64字符串
	 * @param charset 字符集
	 * @return 被加密后的字符串
	 */
	public static String encode(String source, String charset) {
		return encode(StrUtil.bytes(source, charset), charset);
	}
	
	/**
	 * base64编码,URL安全
	 * 
	 * @param source 被编码的base64字符串
	 * @param charset 字符集
	 * @return 被加密后的字符串
	 * @since 3.0.6
	 */
	public static String encodeUrlSafe(String source, String charset) {
		return encodeUrlSafe(StrUtil.bytes(source, charset), charset);
	}

	/**
	 * base64编码
	 * 
	 * @param source 被编码的base64字符串
	 * @param charset 字符集
	 * @return 被加密后的字符串
	 */
	public static String encode(String source, Charset charset) {
		return encode(StrUtil.bytes(source, charset), charset);
	}
	
	/**
	 * base64编码，URL安全的
	 * 
	 * @param source 被编码的base64字符串
	 * @param charset 字符集
	 * @return 被加密后的字符串
	 * @since 3.0.6
	 */
	public static String encodeUrlSafe(String source, Charset charset) {
		return encodeUrlSafe(StrUtil.bytes(source, charset), charset);
	}

	/**
	 * base64编码
	 * 
	 * @param source 被编码的base64字符串
	 * @return 被加密后的字符串
	 */
	public static String encode(byte[] source) {
		return encode(source, CharsetUtil.UTF_8);
	}
	
	/**
	 * base64编码,URL安全的
	 * 
	 * @param source 被编码的base64字符串
	 * @return 被加密后的字符串
	 * @since 3.0.6
	 */
	public static String encodeUrlSafe(byte[] source) {
		return encodeUrlSafe(source, CharsetUtil.UTF_8);
	}

	/**
	 * base64编码
	 * 
	 * @param source 被编码的base64字符串
	 * @param charset 字符集
	 * @return 被加密后的字符串
	 */
	public static String encode(byte[] source, String charset) {
		return StrUtil.str(encode(source, false), charset);
	}
	
	/**
	 * base64编码，URL安全的
	 * 
	 * @param source 被编码的base64字符串
	 * @param charset 字符集
	 * @return 被加密后的字符串
	 * @since 3.0.6
	 */
	public static String encodeUrlSafe(byte[] source, String charset) {
		return StrUtil.str(encodeUrlSafe(source, false), charset);
	}

	/**
	 * base64编码
	 * 
	 * @param source 被编码的base64字符串
	 * @param charset 字符集
	 * @return 被加密后的字符串
	 */
	public static String encode(byte[] source, Charset charset) {
		return StrUtil.str(encode(source, false), charset);
	}
	
	/**
	 * base64编码，URL安全的
	 * 
	 * @param source 被编码的base64字符串
	 * @param charset 字符集
	 * @return 被加密后的字符串
	 * @since 3.0.6
	 */
	public static String encodeUrlSafe(byte[] source, Charset charset) {
		return StrUtil.str(encodeUrlSafe(source, false), charset);
	}
	
	/**
	 * 编码为Base64<br>
	 * 如果isMultiLine为<code>true</code>，则每76个字符一个换行符，否则在一行显示
	 * 
	 * @param arr 被编码的数组
	 * @param isMultiLine 在76个char之后是CRLF还是EOF
	 * @param isUrlSafe 是否使用URL安全字符，一般为<code>false</code>
	 * @return 编码后的bytes
	 */
	public static byte[] encode(byte[] arr, boolean isMultiLine, boolean isUrlSafe) {
		if(null == arr){
			return null;
		}
		
		int len = arr.length;
		if (len == 0) {
			return new byte[0];
		}

		int evenlen = (len / 3) * 3;
		int cnt = ((len - 1) / 3 + 1) << 2;
		int destlen = cnt + (isMultiLine ? (cnt - 1) / 76 << 1 : 0);
		byte[] dest = new byte[destlen];

		byte[] encodeTable = isUrlSafe ? URL_SAFE_ENCODE_TABLE : STANDARD_ENCODE_TABLE;

		for (int s = 0, d = 0, cc = 0; s < evenlen;) {
			int i = (arr[s++] & 0xff) << 16 | (arr[s++] & 0xff) << 8 | (arr[s++] & 0xff);

			dest[d++] = encodeTable[(i >>> 18) & 0x3f];
			dest[d++] = encodeTable[(i >>> 12) & 0x3f];
			dest[d++] = encodeTable[(i >>> 6) & 0x3f];
			dest[d++] = encodeTable[i & 0x3f];

			if (isMultiLine && ++cc == 19 && d < destlen - 2) {
				dest[d++] = '\r';
				dest[d++] = '\n';
				cc = 0;
			}
		}

		int left = len - evenlen;//剩余位数
		if (left > 0) {
			int i = ((arr[evenlen] & 0xff) << 10) | (left == 2 ? ((arr[len - 1] & 0xff) << 2) : 0);

			dest[destlen - 4] = encodeTable[i >> 12];
			dest[destlen - 3] = encodeTable[(i >>> 6) & 0x3f];
			
			if(isUrlSafe){
				//在URL Safe模式下，=为URL中的关键字符，不需要补充。空余的byte位要去掉。
				int urlSafeLen = destlen - 2;
				if(2 == left){
					dest[destlen - 2] = encodeTable[i & 0x3f];
					urlSafeLen += 1;
				}
				byte[] urlSafeDest = new byte[urlSafeLen];
				System.arraycopy(dest, 0, urlSafeDest, 0, urlSafeLen);
				return urlSafeDest;
			}else{
				dest[destlen - 2] = left == 2 ? encodeTable[i & 0x3f] : (byte) '=';
				dest[destlen - 1] = '=';
			}
		}
		return dest;
	}

	// -------------------------------------------------------------------- decode
	/**
	 * base64解码
	 * 
	 * @param source 被解码的base64字符串
	 * @return 被加密后的字符串
	 */
	public static String decodeStr(String source) {
		return decodeStr(source, CharsetUtil.UTF_8);
	}

	/**
	 * base64解码
	 * 
	 * @param source 被解码的base64字符串
	 * @param charset 字符集
	 * @return 被加密后的字符串
	 */
	public static String decodeStr(String source, String charset) {
		return StrUtil.str(decode(source, charset), charset);
	}

	/**
	 * base64解码
	 * 
	 * @param source 被解码的base64字符串
	 * @param charset 字符集
	 * @return 被加密后的字符串
	 */
	public static String decodeStr(String source, Charset charset) {
		return StrUtil.str(decode(source, charset), charset);
	}

	/**
	 * base64解码
	 * 
	 * @param source 被解码的base64字符串
	 * @return 被加密后的字符串
	 */
	public static byte[] decode(String source) {
		return decode(source, CharsetUtil.UTF_8);
	}

	/**
	 * base64解码
	 * 
	 * @param source 被解码的base64字符串
	 * @param charset 字符集
	 * @return 被加密后的字符串
	 */
	public static byte[] decode(String source, String charset) {
		return decode(StrUtil.bytes(source, charset));
	}

	/**
	 * base64解码
	 * 
	 * @param source 被解码的base64字符串
	 * @param charset 字符集
	 * @return 被加密后的字符串
	 */
	public static byte[] decode(String source, Charset charset) {
		return decode(StrUtil.bytes(source, charset));
	}

	/**
	 * 解码Base64，同时解码URL安全和非安全的base64编码
	 * 
	 * @param arr byte数组
	 * @return 解码后的byte数组
	 */
	public static byte[] decode(byte[] arr) {
		if(null == arr){
			return null;
		}
		
		int length = arr.length;
		if (length == 0) {
			return new byte[0];
		}

		int sndx = 0, endx = length - 1;
		int pad = arr[endx] == '=' ? (arr[endx - 1] == '=' ? 2 : 1) : 0;
		int cnt = endx - sndx + 1;
		int sepCnt = length > 76 ? (arr[76] == '\r' ? cnt / 78 : 0) << 1 : 0;
		int len = ((cnt - sepCnt) * 6 >> 3) - pad;
		byte[] dest = new byte[len];

//		byte[] decodeTable = isUrlSafe ? URL_SAFE_DECODE_TABLE : STANDARD_DECODE_TABLE;
		byte[] decodeTable = DECODE_TABLE;

		int d = 0;
		for (int cc = 0, eLen = (len / 3) * 3; d < eLen;) {
			int i = decodeTable[arr[sndx++]] << 18 | decodeTable[arr[sndx++]] << 12 | decodeTable[arr[sndx++]] << 6 | decodeTable[arr[sndx++]];

			dest[d++] = (byte) (i >> 16);
			dest[d++] = (byte) (i >> 8);
			dest[d++] = (byte) i;

			if (sepCnt > 0 && ++cc == 19) {
				sndx += 2;
				cc = 0;
			}
		}

		if (d < len) {
			int i = 0;
			for (int j = 0; sndx <= endx - pad; j++) {
				i |= decodeTable[arr[sndx++]] << (18 - j * 6);
			}
			for (int r = 16; d < len; r -= 8) {
				dest[d++] = (byte) (i >> r);
			}
		}

		return dest;
	}
}
